flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/optimize/QueryHintsResolver.java

```diff
@@ -83,21 +85,50 @@ final List<RelNode> resolve(List<RelNode> roots) {
     }
 
     @Override
-    protected RelNode visitBiRel(BiRel biRel) {
-        Optional<String> leftName = extractAliasOrTableName(biRel.getLeft());
-        Optional<String> rightName = extractAliasOrTableName(biRel.getRight());
-
-        Set<RelHint> existentKVHints = new HashSet<>();
-
-        List<RelHint> oldHints = ((Hintable) biRel).getHints();
+    protected RelNode doVisit(RelNode node) {
+        List<RelHint> oldHints = ((Hintable) node).getHints();
         List<RelHint> oldQueryHints = FlinkHints.getAllQueryHints(oldHints);
         // has no hints, return directly.
         if (oldQueryHints.isEmpty()) {
-            return super.visitChildren(biRel);
+            return super.visitChildren(node);
         }
 
-        List<RelHint> newHints = new ArrayList<>();
+        final List<RelHint> newHints;
+        if (node instanceof BiRel) {
+            BiRel biRel = (BiRel) node;
+            Optional<String> leftName = extractAliasOrTableName(biRel.getLeft());
+            Optional<String> rightName = extractAliasOrTableName(biRel.getRight());
+            newHints = validateAndGetNewHintsFromBiRel(leftName, rightName, oldHints);
+        } else if (node instanceof SingleRel) {
+            SingleRel singleRel = (SingleRel) node;
+            Optional<String> tableName = extractAliasOrTableName(singleRel.getInput());
```

Comment:
Can we get an empty table/alias name here?

Reply:
Emm it seems not, I will modify it separately in this separate jira. WDYT? https://issues.apache.org/jira/browse/FLINK-34235